map2loop/sample_storage.py

```diff
@@ -0,0 +1,268 @@
+            data: The sample data to store.
+        """
+
+        # store the sample data
+        self.samples[sampletype] = data
+        self.sampler_dirtyflags[sampletype] = False
+
+    @beartype.beartype
+    def check_state(self, sampletype: SampleType):
+        """
+        Checks the state of the data, sample and sampler.
+
+        Args:
+            sampletype (SampleType): The type of the sample.
+        """
+
+        self.dirtyflags[StateType.DATA] = self.map_data.dirtyflags[sampletype]
+        self.dirtyflags[StateType.SAMPLER] = self.sampler_dirtyflags[sampletype]
+
+    @beartype.beartype
+    def load(self, sampletype: SampleType):
+        """
+        Loads the map data or raster map data based on the sample type.
+
+        Args:
+            sampletype (SampleType): The type of the sample.
+        """
+        datatype = Datatype(sampletype)
+
+        if datatype == Datatype.DTM:
+            self.map_data.load_raster_map_data(datatype)
+
+        else:
+            # load map data
+            self.map_data.load_map_data(datatype)
+
+    @beartype.beartype
+    def process(self, sampletype: SampleType):
+        """
+        Processes the sample based on the sample type.
+
+        Args:
+            sampletype (SampleType): The type of the sample.
+        """
+
+        if sampletype == SampleType.CONTACT:
+            self.store(
+                SampleType.CONTACT,
+                self.samplers[SampleType.CONTACT].sample(
+                    self.map_data.basal_contacts, self.map_data
+                ),
+            )
+
+        else:
+            datatype = Datatype(sampletype)
+            self.store(
+                sampletype,
+                self.samplers[sampletype].sample(
+                    self.map_data.get_map_data(datatype), self.map_data
+                ),
+            )
+
+    @beartype.beartype
+    def reprocess(self, sampletype: SampleType):
+        """
+        Reprocesses the data based on the sample type.
+
+        Args:
+            sampletype (SampleType): The type of the sample.
+        """
+
+        if sampletype == SampleType.GEOLOGY or sampletype == SampleType.CONTACT:
+            self.map_data.extract_all_contacts()
+
+            if self.project.stratigraphic_column.column is None:
+                self.project.calculate_stratigraphic_order()
+
+            else:
+                self.project.sort_stratigraphic_column()
+
+            self.project.extract_geology_contacts()
+            self.process(SampleType.GEOLOGY)
+
+        elif sampletype == SampleType.STRUCTURE:
+            self.process(SampleType.STRUCTURE)
+
+        elif sampletype == SampleType.FAULT:
+            self.project.calculate_fault_orientations()
+            self.project.summarise_fault_data()
+            self.process(SampleType.FAULT)
+
+        elif sampletype == SampleType.FOLD:
+            self.process(SampleType.FOLD)
+
+    @beartype.beartype
+    def __call__(self, sampletype: SampleType):
```

Comment:
I like having a call method, I think the logic is a bit complicated. It should just be self.process and return the data. All other checks should be done within process.